swarm/src/handler.rs

```diff
@@ -690,6 +724,83 @@ where
     }
 }
 
-/// A statically declared, empty [`HashSet`] allows us to work around borrow-checker rules for
-/// [`ProtocolsAdded::from_set`]. The lifetimes don't work unless we have a [`HashSet`] with a `'static' lifetime.
-static EMPTY_HASHSET: Lazy<HashSet<StreamProtocol>> = Lazy::new(HashSet::new);
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    fn preprocess(s: &'static str) -> HashSet<StreamProtocol> {
+        s.split_whitespace()
+            .map(|p| StreamProtocol::try_from_owned(format!("/{p}")).unwrap())
+            .collect()
+    }
+
+    #[test]
+    fn test_protocol_remove() {
+        let cases = [
+            ["a b c", "a b", "c", "a b"],
+            ["a b c", "a b c", "", "a b c"],
+            ["a b c", "a b c d", "", "a b c"],
+            ["a b c", "d", "a b c", ""],
+            ["", "d", "", ""],
+        ]
+        .map(|arr| arr.map(preprocess));
+
+        for (i, [mut existing, to_remove, reminig, removed]) in cases.into_iter().enumerate() {
+            let mut buffer = Vec::new();
+
+            let change = ProtocolsChange::remove(&mut existing, to_remove, &mut buffer)
+                .into_iter()
+                .flat_map(|c| match c {
+                    ProtocolsChange::Added(_) => panic!("unexpected added"),
+                    ProtocolsChange::Removed(r) => r.cloned(),
+                })
+                .collect::<HashSet<_>>();
+
+            assert_eq!(existing, reminig, "{i}");
+            assert_eq!(change, removed, "{i}");
+        }
+    }
```

Comment:
I am struggling to verify that these tests are actually correct because there is a lot going on. I understand that it will be more code to write but I'd appreciate if we can break this up into more straight-forward tests that have less logic to them and clearly follow the "arrange - act - assert" style.

```rust
// Arrange
let mut existing_protocols = [
	StreamProtocol::new("/foo"),
	StreamProtocol::new("/bar"),
];
let new_protocols = [
	StreamProtocol::new("/bar"),
];

// Act
let changes = change_from_full_sets(existing_protocols, new_protocols, buffer);

// Assert (need some kind of helper to conveniently convert to `Added` and `Removed` variant.
assert_eq!(changes[0], vec![StreamProtocol::new("/bar")]);
```

I don't mind introducing helpers like:

```rust
fn change_from_full_sets(existing: [], new: [], buffer) -> SmallVec<[ProtocolsChange; 2]> {
	....
}
```

What is important to me is that the test is readable and we one can easily cross-check the input state and the assertion.